### PR TITLE
feat(tracer): Record function calls and returns

### PR DESCRIPTION
**Pull request by smanilov**
This commit also records function calls and returns from functions, without passing the actual arguments and values in either direction. That would be a follow-up.

Details:
 * `get_variables` is now public in `tooling/debugger/src/context.rs`; it is needed to extract the stack frame from the debugger;
 * drive-by: `trace_metadata.json` is also stored, because the frontend expects that too;
 * `tail_diff_vecs` is a generic utility function to extract the difference between two vectors; it has good documentation, so details within;
 * a local structure `StackFrame` is introduced, because the `StackFrame` in the debugger does not own its data; here, we need to remember the previous stack trace (a vector of `StackFrame`'s) in order to discover function calls or returns, so we need to own the data;
 * a `stack_trace` field is added to `TracingContext`; see previous point;
 * the new `fn get_stack_trace` performs the conversion mentioned;
 * the old contents of `update_record` are the basis for `tail_diff_vecs`; the new contents use that function to perform the same operation as before; in addition, the core logic of the PR is here: return events are registered for dropped stack frames and call events are registered for new stack frames;
 * all tracing tests are updated, to reflect the new traces that are generated.

To test:

```
cargo test trace
```

### CODE REVIEW

File: tooling/nargo_cli/src/cli/trace_cmd.rs [hunk 1]
```diff
@@ -112,5 +112,11 @@ pub(crate) fn trace_program(
         Err(_) => (),
     }
 
+    let trace_path = Path::new(trace_dir).join("trace_metadata.json");
+    match tracer.store_trace_metadata(&trace_path) {
```
Comment (spartakvarna) [inline tooling/nargo_cli/src/cli/trace_cmd.rs:116]: if you don't care about the Err case then if let is usually enought:
```
if let Ok(_) = tracer.store_trace_metadata(&trace_path) {
    println!("Saved trace to {:?}", trace_path);
}

```

**Review by spartakvarna — APPROVED** [review]

File: tooling/tracer/src/tail_diff_vecs.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+/// assert_eq!(tail_diff_vecs(&vec!(1, 2, 3), &vec!(1, 2)), (2, vec!(3), vec!()));
+///
+/// assert_eq!(tail_diff_vecs(&vec!(1, 2, 3, 4), &vec!(1, 2, 5)), (2, vec!(3, 4), vec!(5)));
+/// ```
+pub fn tail_diff_vecs<T>(a: &Vec<T>, b: &Vec<T>) -> (usize, Vec<T>, Vec<T>)
+where
+    T: PartialEq + Clone,
+{
+    // Here's a functional version of this code, for clarity:
+    // let first_nomatch: usize = xs.iter().zip(ys.iter())
```
Comment (spartakvarna) [inline tooling/tracer/src/tail_diff_vecs.rs:25]: You can short-circuit when first mismatch is found by using the position method of Iter.
```
    let first_nomatch: Option<usize> = a.iter().zip(b.iter())
        .position(|(x, y)| x != y);
```

Comment (smanilov) [reply tooling/tracer/src/tail_diff_vecs.rs]: Thanks, final version is based on your comment.

File: test_programs/plonky2_trace/1_mul/expected_trace.json [hunk 3]
```diff
@@ -3,8 +3,10 @@
   { "Function": { "path_id": 0, "line": -1, "name": "<toplevel>" } },
   { "Call": { "function_id": 0, "args": [] } },
   { "Type": { "kind": 30, "lang_type": "None", "specific_info": { "kind": "None" } } },
-  { "Path": "it-does-not-matter" },
+  { "Path": "<relative-to-this>/src/main.nr" },
```
Comment (EmilIvanichkovv) [inline test_programs/plonky2_trace/1_mul/expected_trace.json:6]: 😎

**Pull request merged**